sparse-array/src/fast.rs

```diff
@@ -0,0 +1,187 @@
+///! Wrapper for sparse arrays that doesn't delete anything from the memory
+/// unless `shrink` is called.
+
+use std::sync::Arc;
+use crate::{
+    bitmap::BitmapIndex,
+    SparseArray,
+};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct FastSparseArray<V> {
+    index: BitmapIndex,
+    data: Arc<SparseArray<V>>,
+}
+
+impl<V: Clone> FastSparseArray<V> {
+    pub fn new() -> Self {
+        Self {
+            index: BitmapIndex::new(),
+            data: Arc::new(SparseArray::new()),
+        }
+    }
+
+    pub fn with_capacity(capacity: u8) -> Self {
+        Self {
+            index: BitmapIndex::new(),
+            data: Arc::new(SparseArray::with_capacity(capacity)),
+        }
+    }
+
+    pub fn get(&self, idx: u8) -> Option<&V> {
+        if !self.index.get_index(idx) {
+            None
+        } else {
+            self.data.get(idx)
+        }
+    }
+
+    pub fn set(&mut self, idx: u8, value: V) {
+        match Arc::get_mut(&mut self.data) {
+            Some(d) => d.set(idx, value),
+            None => {
+                let mut new_sparse_array = SparseArray::new();
+                for (idx, value) in self.iter() {
+                    new_sparse_array.set(idx, (*value).clone());
+                }
+                new_sparse_array.set(idx, value);
+                self.data = Arc::new(new_sparse_array);
+            }
+        }
+        self.index.set_index(idx);
+    }
+
+    pub fn remove(&mut self, idx: u8) -> Option<V> {
+        if self.index.get_index(idx) {
+            self.index.remove_index(idx);
+            return self.data.get(idx).map(|x| (*x).clone());
+        }
+
+        None
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.index.is_empty()
+    }
+
+    pub fn contains_key(&self, idx: u8) -> bool {
+        self.index.get_index(idx)
+    }
+
+    pub fn iter(&self) -> FastSparseArrayIter<V> {
+        FastSparseArrayIter::new(&self)
+    }
+
+    pub fn shrink(&mut self) {
+        let mut new_sparse_array = SparseArray::new();
+        for (idx, value) in self.iter() {
+            new_sparse_array.set(idx, (*value).clone());
+        }
+        self.data = Arc::new(new_sparse_array);
+    }
+}
+
+pub struct FastSparseArrayIter<'a, V> {
+    bitmap: BitmapIndex,
+    sparse_array: &'a FastSparseArray<V>,
+}
+
+impl<'a, V> FastSparseArrayIter<'a, V> {
+    pub fn new(sparse_array: &'a FastSparseArray<V>) -> Self {
+        Self {
+            bitmap: sparse_array.index.clone(),
+            sparse_array,
+        }
+    }
+}
+
+impl<'a, V: Clone> Iterator for FastSparseArrayIter<'a, V> {
```

Comment:
only needed it because `get` requires it too, but same thing it shouldn't be needed